scipy/optimize/_constraints.py

```diff
@@ -0,0 +1,676 @@
+            hess.initialize(len(x0), 'hess')
+
+        # Define function
+        self.f = np.atleast_1d(fun(x0))
+
+        def fun_wrapped(x):
+                return np.atleast_1d(fun(x))
+
+        if jac in ('2-point', '3-point', 'cs'):
+            def fun_wrapped2(x):
+                self.x_diff = x
+                self.f = fun_wrapped(x)
+                return self.f
+        else:
+            fun_wrapped2 = fun_wrapped
+        self.fun = fun_wrapped2
+
+        # Define jacobian
+        if callable(jac):
+            J0 = jac(x0)
+
+            if sparse_jacobian or \
+               (sparse_jacobian is None and sps.issparse(J0)):
+                def jac_wrapped(x):
+                    return sps.csr_matrix(jac(x))
+                self.J = sps.csr_matrix(J0)
+                if isinstance(hess, HessianUpdateStrategy):
+                    self.J_prev = self.J.copy()
+                self.sparse_jacobian = True
+
+            elif sps.issparse(J0):
+                def jac_wrapped(x):
+                    return jac(x).toarray()
+                self.J = J0.toarray()
+                if isinstance(hess, HessianUpdateStrategy):
+                    self.J_prev = np.copy(self.J)
+                self.sparse_jacobian = False
+
+            else:
+                def jac_wrapped(x):
+                    return np.atleast_2d(jac(x))
+                self.J = np.atleast_2d(J0)
+                if isinstance(hess, HessianUpdateStrategy):
+                    self.J_prev = np.copy(self.J)
+                self.sparse_jacobian = False
+
+            if hess in ('2-point', '3-point', 'cs'):
+                def jac_wrapped2(x):
+                    self.x_diff = x
+                    self.J = jac_wrapped(x)
+                    return self.J
+
+            elif isinstance(hess, HessianUpdateStrategy):
+                def jac_wrapped2(x):
+                    self.x_prev = self.x
+                    self.J_prev = self.J
+                    self.x = x
+                    self.J = jac_wrapped(x)
+                    return self.J
+
+            else:
+                jac_wrapped2 = jac_wrapped
+        elif jac in ('2-point', '3-point', 'cs'):
+            J0 = approx_derivative(fun, self.x, f0=self.f,
+                                   **finite_diff_options)
+
+            if sparse_jacobian or \
+               (sparse_jacobian is None and sps.issparse(J0)):
+                def jac_wrapped(x):
+                    J = approx_derivative(fun, self.x, f0=self.f,
+                                          **finite_diff_options)
+                    return sps.csr_matrix(J)
+                self.J = sps.csr_matrix(J0)
+                if isinstance(hess, HessianUpdateStrategy):
+                    self.J_prev = self.J.copy()
+                self.sparse_jacobian = True
+
+            elif sps.issparse(J0):
+                def jac_wrapped(x):
+                    J = approx_derivative(fun, self.x, f0=self.f,
```

Comment:
Just to check: using `self.x` here instead of `x` is not a bug?

Reply:
I have just refactored this.